src/plugins/intellisense/index.tsx

```diff
@@ -0,0 +1,512 @@
+import {
+  PartialFunctionCall,
+  TryFindMQIdentResult,
+  getController,
+  getMathquillIdentifierAtCursorPosition,
+  getPartialFunctionCall,
+} from "./latex-parsing";
+import { IntellisenseState } from "./state";
+import { View, addBracketsToIdent } from "./view";
+import { DCGView, MountedComponent, unmountFromNode } from "DCGView";
+import { MathQuillField, MathQuillView } from "components";
+import { Calc } from "globals/window";
+import { PluginController } from "plugins/PluginController";
+import { getMetadata } from "plugins/manage-metadata/manage";
+import { attach, propGetSet } from "utils/listenerHelpers";
+
+export type BoundIdentifier =
+  | {
+      exprId: string;
+      variableName: string;
+      type:
+        | "variable"
+        | "function-param"
+        | "listcomp-param"
+        | "substitution"
+        | "derivative"
+        | "repeated-operator";
+      id: number;
+    }
+  | BoundIdentifierFunction;
+
+export interface BoundIdentifierFunction {
+  exprId: string;
+  variableName: string;
+  type: "function";
+  id: number;
+  params: string[];
+}
+
+export function getMQCursorPosition(focusedMQ: MathQuillField) {
+  return getController(
+    focusedMQ
+  ).cursor?.cursorElement?.getBoundingClientRect();
+}
+
+const intellisenseMountPoint = document.createElement("div");
+document.body.appendChild(intellisenseMountPoint);
+
+export function getSelectedExpressionID(): string | undefined {
+  return Object.keys(Calc.controller.listModel.selectedItemMap)[0];
+}
+
+export default class Intellisense extends PluginController {
+  static id = "intellisense" as const;
+  static enabledByDefault = true;
+
+  view: MountedComponent | undefined;
+
+  x: number = 0;
+  y: number = 0;
+
+  intellisenseOpts: BoundIdentifier[] = [];
+  intellisenseIndex: number = -1;
```

Comment:
comment: what does `-1` represent? Is it just if the menu isn't open, or any other case?

Reply:
It's that the menu isn't open, though upon further inspection it might be redundant (since I already have `canHaveIntellisense`).

Reply:
Fixed (got rid of this sentinel value)